### PR TITLE
Async API for GAPI : output arguments lifetime tests

### PR DESCRIPTION
**Pull request by anton-potapov**
### This pullrequest add explanatory  test cases for common usage models

- added two template test cases (for calback and future usage model),  emphasising life time of output arguments compared to call to `async`  functions

```
build_gapi_standalone:Linux x64=ade-0.1.1d
build_gapi_standalone:Win64=ade-0.1.1d
build_gapi_standalone:Mac=ade-0.1.1d
build_gapi_standalone:Linux x64 Debug=ade-0.1.1d
```

### CODE REVIEW

**Comment by anton-potapov** [comment]
@andrey-golubev  please take a look

File: modules/gapi/test/gapi_async_test.cpp [hunk 1]
```diff
@@ -380,14 +380,117 @@ TYPED_TEST_P(cancel, basic){
...
+    std::vector<std::shared_ptr<cv::GRunArgs>>    out_ptrs(num_of_requests);
+
+    for (int i=0; i<num_of_requests; i++)
+    {
+        TypeParam r;
+
+        //While input args are copied on on call to async function (actually on call to cv::gin)
+        //and thus do not have to outlive the actual completion of asynchronous activity.
+        //Output arguments are "captured" by reference and therefore _must_ outlive the asynchronous activity
+        //(i.e. live at least until future.get()/wait() is returned)
```
Comment (andrey-golubev) [inline modules/gapi/test/gapi_async_test.cpp]: I feel like this should be added to the API documentation as well. And when it's added there we can assume that user is acquainted with the logic, so a simpler comment would suffice here.

I'd rephrase, let's avoid technical details, the following should work, I suppose:
```
Note: 
Input args are __copied__ and output args are __captured by reference__
User __must__ ensure output arguments are "alive" and accessible from another thread when leaving the current scope
```

Comment (andrey-golubev) [reply modules/gapi/test/gapi_async_test.cpp]: same applies to the above comment in callback version

File: modules/gapi/test/gapi_async_test.cpp [hunk 2]
```diff
@@ -380,14 +380,117 @@ TYPED_TEST_P(cancel, basic){
...
+
 REGISTER_TYPED_TEST_CASE_P(cancel, basic);
 
+template<typename case_t>
+struct output_args_lifetime : ::testing::Test{
+    static constexpr const int num_of_requests = 20;
+};
+TYPED_TEST_CASE_P(output_args_lifetime);
+//There are intentionaly no actual checks (asserts and verify) in output_args_lifetime tests.
+//They are more of example use-cases nature than tests. (ASAN/valgrind can steal catch issues here)
```
Comment (andrey-golubev) [inline modules/gapi/test/gapi_async_test.cpp]: May be:
```
The test cases below describe user scenarios of handling output arguments lifetime when G-API Async API.
These scenarios can be considered examples rather than actual test cases.
```

File: modules/gapi/test/gapi_async_test.cpp [hunk 3]
```diff
@@ -380,14 +380,117 @@ TYPED_TEST_P(cancel, basic){
...
+        auto out_args_ptr =  std::make_shared<cv::GRunArgs>(deep_copy_out_args(r.out_args()));
+
+        //Extend lifetime of out_args_ptr content
+        out_ptrs[i] = out_args_ptr;
+
+        fs[i] = r.async(r.in_args(), args_p_from_args(*out_args_ptr));
+    }
+
+    for (auto&& ftr : fs ){
+        ftr.get();
```
Comment (andrey-golubev) [inline modules/gapi/test/gapi_async_test.cpp]: I wonder if `ftr.wait()` is better here since we're not interested in the result itself

File: modules/gapi/test/gapi_async_test.cpp [hunk 4]
```diff
@@ -380,14 +380,117 @@ TYPED_TEST_P(cancel, basic){
...
+        //(i.e. live at least until future.get()/wait() is returned)
+        auto out_args_ptr =  std::make_shared<cv::GRunArgs>(deep_copy_out_args(r.out_args()));
+
+        //Extend lifetime of out_args_ptr content
+        out_ptrs[i] = out_args_ptr;
+
+        fs[i] = r.async(r.in_args(), args_p_from_args(*out_args_ptr));
+    }
+
+    for (auto&& ftr : fs ){
```
Comment (andrey-golubev) [inline modules/gapi/test/gapi_async_test.cpp]: why `auto&&` not `auto&`?

File: modules/gapi/test/gapi_async_test.cpp [hunk 5]
```diff
@@ -380,14 +380,117 @@ TYPED_TEST_P(cancel, basic){
...
+    {
+        TypeParam r;
+
+        //While input args are copied on on call to async function (actually on call to cv::gin)
+        //and thus do not have to outlive the actual completion of asynchronous activity.
+        //Output arguments are "captured" by reference and therefore _must_ outlive the asynchronous activity
+        //(i.e. live at least until callback is called)
+        auto out_args_ptr =  std::make_shared<cv::GRunArgs>(deep_copy_out_args(r.out_args()));
+
+        //Extend lifetime of out_mat_ptr content by capturing it into a callback
```
[the remark this answers is not in the document]
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: thanks - fixed

File: modules/gapi/test/gapi_async_test.cpp (hunk 3 above)
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: fixed

File: modules/gapi/test/gapi_async_test.cpp (hunk 4 above)
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: changed to `const auto&`

File: modules/gapi/test/gapi_async_test.cpp [hunk 6]
```diff
@@ -380,14 +380,117 @@ TYPED_TEST_P(cancel, basic){
...
+
+    GRunArgsP args_p_from_args(GRunArgs& args){
+        GRunArgsP result; result.reserve(args.size());
+        for (auto&& arg : args){
+            switch (arg.index()){
+    #if !defined(GAPI_STANDALONE)
+                case GRunArg::index_of<cv::Mat>()                 :   result.emplace_back(&util::get<cv::Mat>(arg));     break;
+                case GRunArg::index_of<cv::Scalar>()              :   result.emplace_back(&util::get<cv::Scalar>(arg));  break;
+                case GRunArg::index_of<cv::UMat>()                :   result.emplace_back(&util::get<cv::UMat>(arg));    break;
+    #endif // !defined(GAPI_STANDALONE)
```
[the remark this answers is not in the document]
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: this `switch` is poor-man `std::visit`  for `std::variant`. So it should handle all possible cases of `variant` content

File: modules/gapi/test/gapi_async_test.cpp (hunk 1 above)
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: changed

File: modules/gapi/test/gapi_async_test.cpp (hunk 2 above)
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: I would prefer to keep existing description

Comment (andrey-golubev) [reply modules/gapi/test/gapi_async_test.cpp]: okay

File: modules/gapi/test/gapi_async_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrey-golubev) [reply modules/gapi/test/gapi_async_test.cpp]: hmm, I don't see it as fixed, are you sure?

**Comment by andrey-golubev** [comment]
@anton-potapov hmm, I see you commented "fixed" (and similar) in some places but I don't actually see any change. may be you forgot to push the updates?

File: modules/gapi/test/gapi_async_test.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrey-golubev) [reply modules/gapi/test/gapi_async_test.cpp]: hmm, okay then. consider resolved

File: modules/gapi/test/gapi_async_test.cpp (hunk 1 above)
Comment (andrey-golubev) [reply modules/gapi/test/gapi_async_test.cpp]: good, consider resolved

File: modules/gapi/test/gapi_async_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrey-golubev) [reply modules/gapi/test/gapi_async_test.cpp]: now I do, consider resolved

File: modules/gapi/test/gapi_async_test.cpp (hunk 3 above)
Comment (andrey-golubev) [reply modules/gapi/test/gapi_async_test.cpp]: thanks, consider resolved

File: modules/gapi/test/gapi_async_test.cpp (hunk 4 above)
Comment (andrey-golubev) [reply modules/gapi/test/gapi_async_test.cpp]: consider resolved

**Comment by anton-potapov** [comment]
@dmatveev  please take a look

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp [hunk 7]
```diff
@@ -24,6 +24,11 @@ namespace wip {
     //These functions asynchronously (i.e. probably on a separate thread of execution) call operator() member function of their first argument with copies of rest of arguments (except callback) passed in.
     //The difference between the function is the way to get the completion notification (via callback or a waiting on std::future object)
     //If exception is occurred during execution of apply it is transfered to the callback (via function parameter) or passed to future (and will be thrown on call to std::future::get)
+
+    //N.B. :
+    //Input arguments are copied on on call to async function (actually on call to cv::gin) and thus do not have to outlive the actual completion of asynchronous activity.
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: on on

File: modules/gapi/include/opencv2/gapi/gcomputation_async.hpp [hunk 8]
```diff
@@ -25,6 +25,11 @@ namespace wip  {
     //These functions asynchronously (i.e. probably on a separate thread of execution) call apply member function of their first argument with copies of rest of arguments (except callback) passed in.
     //The difference between the function is the way to get the completion notification (via callback or a waiting on std::future object)
     //If exception is occurred during execution of apply it is transfered to the callback (via function parameter) or passed to future (and will be thrown on call to std::future::get)
+
+    //N.B. :
+    //Input arguments are copied on on call to async function (actually on call to cv::gin) and thus do not have to outlive the actual completion of asynchronous activity.
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gcomputation_async.hpp]: Here too

File: modules/gapi/test/gapi_async_test.cpp [hunk 9]
```diff
@@ -380,14 +380,115 @@ TYPED_TEST_P(cancel, basic){
...
+
 REGISTER_TYPED_TEST_CASE_P(cancel, basic);
 
+template<typename case_t>
+struct output_args_lifetime : ::testing::Test{
+    static constexpr const int num_of_requests = 20;
+};
+TYPED_TEST_CASE_P(output_args_lifetime);
+//There are intentionaly no actual checks (asserts and verify) in output_args_lifetime tests.
+//They are more of example use-cases than real tests. (ASAN/valgrind can steal catch issues here)
```
Comment (dmatveev) [inline modules/gapi/test/gapi_async_test.cpp]: steal -> still

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp (hunk 7 above)
Comment (anton-potapov) [reply modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: fixed

File: modules/gapi/include/opencv2/gapi/gcomputation_async.hpp (hunk 8 above)
Comment (anton-potapov) [reply modules/gapi/include/opencv2/gapi/gcomputation_async.hpp]: fixee

File: modules/gapi/test/gapi_async_test.cpp (hunk 9 above)
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: fixed

**Comment by dmatveev** [comment]
Thanks!

**Comment by dmatveev** [comment]
Not sure if the standalone build worked, though (due to PR formatting?)
